New session handling

nsklikas opened the pull request:
@rohe mentioned that the new session management system is ready and that we should start discussing whether it should be merged and if any changes are needed.

So I am taking the liberty to create this PR to discuss any changes we see necessary.

rohe committed:
Handle different types of input.
There might not be any client information at this time. This is an effect of automatic client registration as defined in the OIDC federation specification.

rohe committed:
Handle different types of input.
There might not be any client information at this time. This is an effect of automatic client registration as defined in the OIDC federation specification.

rohe committed:
Allow for adding extra response arguments. Not used here but used in classes that are built on this.

rohe committed:
Allow for adding extra response arguments. Not used here but used in classes that are built on this.

rohe committed:
Bumped version.

rohe committed:
Bumped version.

rohe committed:
clean up

rohe committed:
clean up

rohe committed:
Change so the method definition is the same.

rohe committed:
Change so the method definition is the same.

rohe committed:
Basic functionality.

rohe committed:
Basic functionality.

rohe committed:
Fixed tests.

rohe committed:
Fixed tests.

rohe committed:
Refactoring and adding more usecases.

rohe committed:
Refactoring and adding more usecases.

rohe committed:
Deal with commonality.

rohe committed:
Deal with commonality.

rohe committed:
Merge branch 'new_session_handling' of github.com:IdentityPython/oidcendpoint into new_session_handling

rohe committed:
Getting closer

rohe committed:
Getting closer

rohe committed:
A couple more done.

rohe committed:
A couple more done.

rohe committed:
All functionality there and all tests working!
Now, for documentation.

rohe committed:
All functionality there and all tests working!
Now, for documentation.

rohe committed:
Corrected attribute name.

rohe committed:
Corrected attribute name.

rohe committed:
Started on documentation.

rohe committed:
Started on documentation.

rohe committed:
Add session management documentation.

rohe committed:
Add session management documentation.

rohe committed:
Add session management documentation.

rohe committed:
Add session management documentation.

rohe committed:
Spelling error

rohe committed:
Spelling error

rohe committed:
Avoid name clash.

rohe committed:
Avoid name clash.

rohe committed:
Better function names. session_key instead of db_key.

rohe committed:
Better function names. session_key instead of db_key.

rohe committed:
Get rid of salt from AuthnEvent, was not used.

rohe committed:
Get rid of salt from AuthnEvent, was not used.

rohe committed:
isort
Removed duplicated code in oauth2/oidc Authorization classes.

rohe committed:
isort
Removed duplicated code in oauth2/oidc Authorization classes.

rohe committed:
Make the oauth2 Authorization class the base class.

rohe committed:
Make the oauth2 Authorization class the base class.

rohe committed:
The common module not needed anymore.

rohe committed:
The common module not needed anymore.

rohe committed:
Added tests.
A token must have a value.

rohe committed:
Added tests.
A token must have a value.

rohe committed:
Clarifications and more tests.

rohe committed:
Clarifications and more tests.

rohe committed:
More tests.

rohe committed:
More tests.

rohe committed:
Switched to using Message as base for SessionInfo.

rohe committed:
Switched to using Message as base for SessionInfo.

rohe committed:
Deal with custom token maps.

rohe committed:
Deal with custom token maps.

rohe committed:
Parameter name change.
Added types.

rohe committed:
Parameter name change.
Added types.

rohe committed:
new test - test_get_spec

rohe committed:
new test - test_get_spec

rohe committed:
If authentication event has timed out
do new login.

rohe committed:
If authentication event has timed out
do new login.

rohe committed:
Simplified interface does not support the replace and update methods anymore.

rohe committed:
Simplified interface does not support the replace and update methods anymore.

rohe committed:
Removed unnecessary argument.

rohe committed:
Removed unnecessary argument.

rohe committed:
Added useful method get_claims_all_usage.

rohe committed:
Added useful method get_claims_all_usage.

rohe committed:
JWTToken grant management aware.
More type hints.
'token' another user info interface handled by the claims interface.

rohe committed:
JWTToken grant management aware.
More type hints.
'token' another user info interface handled by the claims interface.

rohe committed:
Refactored

rohe committed:
Refactored

rohe committed:
isort

rohe committed:
isort

rohe committed:
More type hints.
Changed parameters to attributes to make Item and Grant similar

rohe committed:
More type hints.
Changed parameters to attributes to make Item and Grant similar

rohe committed:
Client_verification_method is better to use in the OIDC fed context.
More unit tests.

rohe committed:
Client_verification_method is better to use in the OIDC fed context.
More unit tests.

rohe committed:
A simple ':' as divider does not work.

rohe committed:
A simple ':' as divider does not work.

rohe commented:
Thanks Nikos !

nsklikas commented:
Is it ok if I rebase this to latest to fix the conflicts?

peppelinux commented:
Hi @nsklikas, great to know that you walked over this, finally!
I'll be quite busy for a while until Gen 2021 and I'd need to release djangosaml2 v1.0.0 first of all, so I'm keeping an eye on this PR but feel free to overhelm my review if that could not come in time.

rohe commented:
Absolutely OK to rebase.

nsklikas commented on src/oidcendpoint/authn_event.py:
```diff
@@ -24,30 +25,36 @@ def expires_in(self):
     """
 
+    :param authn_time:
     :param uid:
-    :param salt:
     :param authn_info:
+    :param valid_until:
+    :param expires_in:
     :param kwargs:
```
Fix params, not `uid` param. Move `authn_info` to start:
```
:param authn_info
:param authn_time
:param valid_until
:param expires_in
```

rohe replied:
I think this is already in there. At least I have it in my copy. So no problem with this.

rohe replied:
UID is used further down. Why do you say it's not  needed ?

nsklikas commented on src/oidcendpoint/authn_event.py:
```diff
@@ -24,30 +25,36 @@ def expires_in(self):
+    if authn_time:
+        args["authn_time"] = authn_time
+    else:
+        _ts = kwargs.get("timestamp")
```
Let's add `timestamp` argument to function. We shouldn't use undocumented arguments from kwargs. This way the caller knows all the params he can pass, without having to look into the function body

peppelinux replied:
looks good, is there some exception handling if "timestamp" couldn't be available?

rohe replied:
OK

peppelinux replied:
what would be a default value on `_ts`?

nsklikas commented on src/oidcendpoint/authz/__init__.py:
```diff
@@ -14,50 +17,43 @@ class AuthzHandling(object):
-
-    def permissions(self, cookie=None, **kwargs):
-        return self.permission
+    def __init__(self, endpoint_context, **kwargs):
+        AuthzHandling.__init__(self, endpoint_context, **kwargs)
```
Not needed

rohe replied:
Matter of programming style. But yes, not needed.

nsklikas commented on src/oidcendpoint/endpoint.py:
```diff
@@ -197,6 +198,7 @@ def __init__(self, endpoint_context, **kwargs):
         # This is for matching against aud in JWTs
         # By default the endpoint's endpoint URL is an allowed target
         self.allowed_targets = [self.name]
+        self.client_verification_method = []
```
This is unused

peppelinux replied:
we could execute autopep8 and autoflake on this branch

rohe replied:
It's needed by OIDC federation.

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+                            _gid_key = session_key(uid, client_id, grant_id)
+                            if grant_id in _cid_info['subordinate']:
+                                _gid_info = self._db[_gid_key]
+                                if not _gid_info:
+                                    self._db[_cid_key] = _cid_info.add_subordinate(grant_id)
```
Isn't the `grant_id` already a `_cid_info subordinate`. It was checked in https://github.com/IdentityPython/oidcendpoint/pull/104/files#diff-93bc9f688c9bce53870f2c48de99c5e6b094668e243a710ce6d72abad9708a79R54

rohe replied:
Not clear which line/-s you're referring to.
Remember there are 2 places subordinates appear:
1) as a identifier in the list of subordinates to an object. (for example _cid_info['subordinate'])
2) by value in the database (for example self._db[_gid_key])

So, in order to add a subordinate you have to 
1) Add the identifier to the 'subordinate' parameter and
2) Add the information connected to the subordinate in the database

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+                            self._db[_cid_key] = value
+                    else:
+                        _userinfo.add_subordinate(client_id)
+                        if grant_id:
+                            _cid_info = ClientSessionInfo()
+                            _cid_info.add_subordinate(grant_id)
+                            self._db[_cid_key] = _cid_info
+                            self._db[session_key(uid, client_id, grant_id)] = value
+                        else:
+                            _cid_info = ClientSessionInfo()
+                            self._db[_cid_key] = _cid_info
```
I think this would be cleaner:
```python
_cid_info = ClientSessionInfo()
if grant_id:
    _cid_info.add_subordinate(grant_id)
    self._db[session_key(uid, client_id, grant_id)] = value
self._db[_cid_key] = _cid_info
```

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+                            self._db[_cid_key] = _cid_info
+                            self._db[session_key(uid, client_id, grant_id)] = value
+                        else:
+                            _cid_info = ClientSessionInfo()
+                            self._db[_cid_key] = _cid_info
```
Shouldn't this be:
```python
self._db[_cid_key] = value
```

rohe replied:
Right! Need to add a set of tests. There was none covering this case.

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+                    else:
+                        _userinfo.add_subordinate(client_id)
+                        if grant_id:
+                            _cid_info = ClientSessionInfo()
+                            _cid_info.add_subordinate(grant_id)
+                            self._db[_cid_key] = _cid_info
+                            self._db[session_key(uid, client_id, grant_id)] = value
+                        else:
+                            _cid_info = ClientSessionInfo()
+                            self._db[_cid_key] = _cid_info
+                        self._db[uid] = _userinfo
+                else:
+                    _userinfo.add_subordinate(client_id)
+                    self._db[uid] = _userinfo
+                    if grant_id:
+                        _cid_info = ClientSessionInfo()
+                        _cid_info.add_subordinate(grant_id)
+                        self._db[session_key(uid, client_id, grant_id)] = value
+                    else:
+                        _cid_info = value
+
+                    _cid_key = session_key(uid, client_id)
+                    self._db[_cid_key] = _cid_info
+            else:
+                self._db[uid] = value
+        else:
+            if client_id:
+                _user_info = UserSessionInfo()
+                _user_info.add_subordinate(client_id)
+                if grant_id:
+                    _cid_info = ClientSessionInfo()
+                    _cid_info.add_subordinate(grant_id)
+                    self._db[session_key(uid, client_id, grant_id)] = value
+                else:
+                    _cid_info = value
+                self._db[session_key(uid, client_id)] = _cid_info
+            else:
+                _user_info = value
+
+            self._db[uid] = _user_info
```
I think this can be refactored to:
```python
        uid, client_id, grant_id = self._eval_path(path)

        if grant_id:
            gid_key = session_key(uid, client_id, grant_id)
            self._db[gid_key] = value

        if client_id:
            cid_key = session_key(uid, client_id)
            if not grant_id:
                self._db[cid_key] = value
            elif grant_id not in cid_info["subordinate"]:
                cid_info = self._db.get(cid_key, ClientSessionInfo())
                cid_info.add_subordinate(grant_id)
                self._db[cid_key] = cid_info

        userinfo = self._db.get(uid, UserSessionInfo())
        if not client_id:
            self._db[uid] = value
        if client_id not in userinfo["subordinate"]:
            userinfo.add_subordinate(client_id)
            self._db[uid] = userinfo
```

rohe replied:
Nope!
Tested with the extended set of tests and a bunch failed.
Still it might be worth seeing if your refactoring only needs a small set of tweaks to work.

rohe replied:
Just needed a few modifications.

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+            user_info = self._db[uid]
+        except KeyError:
+            raise KeyError('No such UserID')
+        else:
```
This `else` is redundant

rohe replied:
Not relevant anymore

nsklikas commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,173 @@
+                    client_session_info = self._db[session_key(uid, client_id)]
+                except KeyError:
+                    return SessionInfo()
+                else:
```
This `else` is redundant

rohe replied:
Not relevant anymore

nsklikas commented on src/oidcendpoint/session/manager.py:
```diff
@@ -0,0 +1,290 @@
+        args = {k: v for k, v in kwargs.items() if k in Grant.attributes}
+        _grant = Grant(**args)
+        self.set([user_id, client_id, _grant.id], _grant)
+        _client_session_info = self.get([user_id, client_id])
+        _client_session_info["subordinate"].append(_grant.id)
+        self.set([user_id, client_id], _client_session_info)
```
This is done by `Database.set` (https://github.com/IdentityPython/oidcendpoint/pull/104/files#diff-93bc9f688c9bce53870f2c48de99c5e6b094668e243a710ce6d72abad9708a79R60)

rohe replied:
So your proposal is to remove lines 256-257 ?

nsklikas replied:
257-258

nsklikas commented:
Maybe I don't understand the logic behind it, but it doesn't make sense to me where the authentication event and the authentication request are stored.

The authentication event is stored in the `UserSessionInfo` and the authentication request is stored in the `ClientSessionInfo`. But what happens if a user logins many times? The `authn_time` and `valid_until` would be different, so we would have to store each event. It seems to me that the usersession won't be updated (https://github.com/IdentityPython/oidcendpoint/pull/104/files#diff-3dc0019abb463dacf7420fd38473e3172fa3ed7ec5e7adb87a03d50feee42860R102-R106). Also the same user may login many times with the same client, this means that a (user_id, client_id) pair may correspond to many different authentication requests. 

I think that it would make more sense if we stored the `authentication_event` and `authentication_request` in the `Grant`. This would mean that we wouldn't be able to have 1 user authentication shared across many clients. I'm not sure how we should move forward if we would like to support this

nsklikas commented on src/oidcendpoint/oauth2/authorization.py:
```diff
@@ -100,27 +290,57 @@ class Authorization(Endpoint):
+        token = grant.mint_token(
+            token_type,
+            value=_mngr.token_handler["access_token"](
+                session_info["session_id"],
+                client_id=session_info["client_id"],
+                aud=grant.resources,
+                user_claims=None,
+                scope=grant.scope,
+                sub=session_info["client_session_info"]['sub']
```
We shouldn't require the whole `session_info` if we only need the `client_session_info`

rohe replied:
Also needs session_id. Does it hurt ?

nsklikas commented on src/oidcendpoint/oauth2/introspection.py:
```diff
@@ -23,59 +22,37 @@ class Introspection(Endpoint):
+            "active": True,
+            "scope": " ".join(scope),
+            "client_id": session_info["client_id"],
+            "token_type": token.type,
```
`token_type` is not supposed to be `access_token` or `refresh_token`. It should be `bearer` or `mac` (https://tools.ietf.org/html/rfc6749#section-7.1)

rohe replied:
OK! Presently we're only dealing with `bearer` type tokens.

nsklikas commented on src/oidcendpoint/oidc/add_on/pkce.py:
```diff
@@ -90,12 +90,13 @@ def post_token_parse(request, client_id, endpoint_context, **kwargs):
 
     try:
-        _info = endpoint_context.sdb[request["code"]]
+        _session_info = endpoint_context.session_manager.get_session_info_by_token(request["code"])
```
Why retrieve the whole session_info if all we want is the `client_session_info`

rohe replied:
Sure we can add another method don't see that we gain a lot by doing it though.

nsklikas replied:
Fewer db queries, why do 3 queries to retrieve the user session, the client session and the grant session when you only want the client session

peppelinux commented:
I read your comments and looks good, probably Roland knows why something was designed in that way, at the same time a lone developer have to go ahead and then we'll see. Now we see and the less code with more simplicity, with same features and behaviour at the same time, would be the best way.

I'm still not get completely in, in the new session storage. I'd need some weeks to get something up, concretely. If you can go ahead It would be good to me, if our oidc group would have a meeting to see the next roadmap and your comments It would be great

rohe commented:
One of the basic assumptions I made when writing this was that we want to support Single Sign On (SSO) and Single Logout where possible. That means that one authentication by a user MUST be valid for all clients that user uses. That's the whole definition of SSO in my view. Of course when the authentication times out so does all the sign ins from all clients that used that authentication. 
Support SSO means that the authentication is user specific NOT client specific, hence it made sense to me to store it in the User Session Info.
If you don't want to support SSO then having the authentication event among other Client session info makes sense but that use case was not what I was aiming for.

peppelinux commented:
> Support SSO means that the authentication is user specific NOT client specific, hence it made sense to me to store it in the User Session Info.

Thank you Roland, that's the interpratation of the use case!
@nsklikas probably would like to introduce another relational model between these two entities, how that would be?
Interesting

rohe commented:
Thought about that and I would make it configurable and add an authentication information interface that would go looking in different place depending on whether SSO or no-SSO was configured.

rohe committed:
Increased test coverage of session/database.py to 100%.

rohe committed:
Increased test coverage of session/database.py to 100%.

rohe committed:
Refactored get() based on nsklikas proposal.

rohe committed:
Refactored get() based on nsklikas proposal.

nsklikas committed:
Fix rebase

Some tests were commented out because they used token_coop instead of
token which doesn't support the same functionality.

nsklikas commented:
I rebased on develop. I had to comment out some tests because token_coop was removed and token doesn't have the same functionality. Perhaps we should implement the missing functionality on token?

Regarding the database. Yes, I understand that you designed it with SSO in mind. But even in that case we must be able to have multiple authentication events per user, e.g. a user might try to log in from different browsers resulting in 2 active authentication events. 
In addition one authentication event will eventually time out. With the current implementation (I think I haven't tested it yet) the new authentication event will never overwrite the old one. But even overwriting the old one might not be correct in all cases, as some users will probably want to maintain a log of all logins.

Aside from these comments IMHO we should do something like you describe and support both use cases. I don't think whether SSO is supported or not should be reflected on the database schema (it should be implemented on a software level).

nsklikas commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -54,21 +94,16 @@ def _access_token(self, req, **kwargs):
-            return self.error_cls(
-                error="invalid_grant", error_description="Session is revoked"
-            )
+        if "state" in req:
+            # verify that state in this request is the same as the one in the
+            # authorization request
+            if _authn_req["state"] != req["state"]:
```
What if `state` is defined in `_authn_req` but not in `req`? Then we won't throw an error.

Also what if `state` is not defined in `_authn_req` but it is in `req`? A KeyError will be raised.

We should write tests for these cases. I can do that if you want

nsklikas commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -80,29 +115,46 @@ def _access_token(self, req, **kwargs):
+            "token_type": "Bearer",
+            "expires_in": 900,
+            "scope": grant.scope,
+            "state": _authn_req["state"]
```
`state` may not be present in the original `_auth_req`

nsklikas replied:
Bump

rohe committed:
Use the parents __init__ .

nsklikas commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -111,49 +150,144 @@ def _access_token(self, req, **kwargs):
+        _resp = {
+            "access_token": access_token.value,
+            "token_type": "Bearer",
+            "expires_in": 900,
```
Why is this hard coded?

rohe replied:
Fixed.

nsklikas commented on src/oidcendpoint/session/__init__.py:
```diff
@@ -0,0 +1,17 @@
+
+
+def session_key(*args) -> str:
+    return ';;'.join(args)
```
Move `;;` to variable

rohe replied:
Done

nsklikas commented on src/oidcendpoint/session/manager.py:
```diff
@@ -0,0 +1,290 @@
+    @staticmethod
+    def _pick_grant(client_info, token_type):
+        if len(client_info["subordinate"]) == 1:
+            return client_info["subordinate"][0]
```
Shouldn't we match with the `token_type` here?

nsklikas commented on src/oidcendpoint/oauth2/authorization.py:
```diff
@@ -528,79 +795,82 @@ def post_authentication(self, user, request, sid, **kwargs):
+
+        client_info = ClientSessionInfo(
+            authorization_request=request,
+            sub=_mngr.sub_func['public'](user_id, salt=_mngr.salt)
```
The sub_func used should be configurable

nsklikas commented on src/oidcendpoint/oidc/authorization.py:
```diff
@@ -123,601 +98,41 @@ class Authorization(Endpoint):
-            else:
-                _context.sdb.update(sid, code=None)
-                _code = None
+    def setup_client_session(self, user_id: str, request: dict) -> str:
```
Perhaps we should allow the caller to pass his own `sector_identifier_uri` and `salt` to allow for greater flexibility

rohe replied:
`sector_identifier_uri` comes from the authorization request and should not be entered any other way.

rohe committed:
Changed get_session_info(), get_session_info_by_token() APIs such that you can chose what to get from the session DB.

rohe committed:
Changed get_session_info(), get_session_info_by_token() APIs such that you can chose what to get from the session DB.

rohe committed:
Ran isort.

rohe committed:
Merge branch 'new_session_handling' of github.com:IdentityPython/oidcendpoint into new_session_handling

rohe commented:
OK, I think I know how to deal with supporting SSO and having a user authenticating with more then one computer/browser.
Will verify !

nsklikas commented on src/oidcendpoint/token/jwt_token.py:
```diff
@@ -0,0 +1,123 @@
+        self.def_aud = aud or []
+        self.alg = alg
+
+    def __call__(self, session_id: str, **kwargs):
+        """
+        Return a token.
+
+        :param endpoint_context:
+        :param session_id: Session id
+        :return: Signed JSON Web Token
+        """
+
+        session_info = self.endpoint_context.session_manager.get_session_info(
+            session_id, client_session_info=True, grant=True)
+        sub = session_info["client_session_info"]["sub"]
+        payload = {"sid": session_id, "ttype": self.type, "sub": sub}
+
+        grant = session_info["grant"]
+        _claims_restriction = grant.claims.get("token")
+        if _claims_restriction:
+            user_info = self.endpoint_context.claims_interface.get_user_claims(
+                session_info["user_id"], _claims_restriction)
+            payload.update(user_info)
+
+        # payload.update(kwargs)
+        signer = JWT(
+            key_jar=self.key_jar,
+            iss=self.issuer,
+            lifetime=self.lifetime,
+            sign_alg=self.alg,
+        )
+
+        _aud = kwargs.get('aud')
+        if _aud is None:
+            _aud = grant.resources
+        else:
+            _aud = _aud if isinstance(_aud, list) else [_aud]
+            _aud.extend(grant.resources)
+
+        return signer.pack(payload, aud=_aud)
```
The `add_scope` option was removed. Is there another way to add the scopes to the JWT or was the feature abandoned altogether?

rohe replied:
Granted this is an omission but adding scopes should not be done only here but also when using mint_token()
because the Token instance must contain the information that is put in the JWT Token.
The JWT Token MUST be based on what's in the Token. Have to make sure of the order.

nsklikas commented on src/oidcendpoint/oidc/userinfo.py:
```diff
@@ -96,23 +94,26 @@ def do_response(self, response_args=None, request=None, client_id="", **kwargs):
+                                                        user_session_info=True,
+                                                        client_session_info=True,
+                                                        grant=True)
+        token = _mngr.find_token(_session_info["session_id"], request["access_token"])
```
We don't know if this really is an access token, it could also be a refresh token (I have tested this).

This probably should be changed in `parse_request` https://github.com/IdentityPython/oidcendpoint/blob/27a0a4d27d9e53c4fb1e3bfd75dfb0a644bf7736/src/oidcendpoint/oidc/userinfo.py#L164

rohe replied:
Added tests to verify right kind of token.

nsklikas commented:
One other thing that concerns me is the claim management. The `ClaimsInterface.get_claims` function is used to decide the claims each endpoint can return. The problem to me is that we are changing the behaviour the library used to have and also that the new behaviour seems counter intuitive to me.  
The new behaviour is that a set of claims is collected per endpoint using:
- Default claims (`base_claims`)
- Per client claims (`{}_claims_per_client`)
- Scopes are mapped to claims (`add_claims_by_scope`)

This set of claims is returned if not claims have been requested from the client (using a claims request). But if a claims request was made than the set of claims we described earlier is used to limit the claims requested. IMO this causes several problems, e.g. we can't add certain claims to all idtokens (because if the client requests extra claims only those will be returned). Also the client can't request claims from the userinfo endpoint using both scopes and claims request, because then only the claims defined in the claims request will be returned and the scopes will only be used to limit them. 

The spec says (https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter):
```
userinfo
    OPTIONAL. Requests that the listed individual Claims be returned from the UserInfo Endpoint. If present, the listed Claims are being requested to be added to any Claims that are being requested using scope values. If not present, the Claims being requested from the UserInfo Endpoint are only those requested using scope values. 
```

I think that the behaviour of the claims interface should be:
- Collect the claims set the same way it is done now.
- Limit the requested claims using a (per client?) configuration, e.g. `allowed_claims` (or `allowed_requested_claims`)
- Return the union of these to sets.

rohe committed:
Rewritten to deal with authorization requests using different browsers and/or different content.

rohe committed:
Tests for wrong type of token and invalid token.

rohe committed:
isort

rohe commented:
I actually think that mapping scopes to claims and specifying claims using the claims parameter are alternative ways of requesting claims. They should NOT be used together. If they are, whatever is defined using the claims parameter overrides the scope->claims transformation.

rohe committed:
Don't hardcode.

angelakis commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -83,26 +118,91 @@ def _access_token(self, req, **kwargs):
+                                             code)
+            _response["refresh_token"] = refresh_token.value
+
+        code.register_usage()
```
This, unfortunately, does not work if the session db is not an in-memory dict. Every attribute change should use the get/set database interface.

rohe replied:
Fix been pushed.

angelakis replied:
Thanks!

angelakis replied:
Unfortunately, there's still an issue here when using a database. The grant that is updated does not contain the code object that just got changed (with `register_usage`). The reason is the methods in line 92 and 95 (`get_session_info_by_token` and `find_token`).

More specifically, `get_session_info_by_token` fetches the whole session_info (including the issued_tokens). After that instead of looking into the fetched objects and finding the right code, we use `find_token` which AGAIN looks into the database and finds the correct token. As a result, the same code is fetched but now it is a different object because the `to_json` method of course creates a new one. 

I recommend we change the `find_token` method or create a new one that instead of querying the database, is given a session_info and tries to find a token inside the object. That way we both solve our bug and reduce the queries needed in the database.

nsklikas replied:
The `grant` in `find_token` could be an optional param, when that is `None` we query the database:
```python
    def find_token(self, token_value: str, session_id: str = None, grant: Grant = None) -> Optional[Token]:
        """
        Retrieve a token from a grant. If a grant object is not be provided the session_id is used to 
        retrieve it from the database. At least one of grant and session_id must be provided.

        :param token_value:
        :param session_id: Based on 3-tuple, user_id, client_id and grant_id
        :param grant: A grant object
        :return:
        """
        if not grant:
            if not session_id:
                raise ValueError("One of `grant` or `session_id` must be provided")
            user_id, client_id, grant_id = unpack_session_key(session_id)
            grant = self.get([user_id, client_id, grant_id])
        for token in grant.issued_token:
            if token.value == token_value:
                return token

        return None
```

angelakis replied:
Unfortuanately, this still does not work for databases.

In line 92 `get_session_info_by_token`  is called so the session_info is fetched from the DB and created with `from_json`. After that, `find_token` is called in line 95 and the database is again queried (although we already have the token inside session_info's issued_tokens). As a result, the code returned is again created with `from_json` and although it's the same as the one inside session_info, it's a different object. 

So when `code.register_usage()` is called the code object changes (self.used is incremented) but the code inside `session_info` stays unchanged.

This could be solved if instead of using `find_token` (and querying the database a second time without reason) we could fetch the code from inside `session_info` 

         _session_info = _mngr.get_session_info_by_token(_access_code, grant=True)
        code = _mngr.find_token(_session_info["session_id"], _access_code)

angelakis replied:
Mentioning @rohe because this conversation has been lost after some pushes.

rohe committed:
Move creation of the string representation of a token to happen in the context of the token.

rohe committed:
To deal with the case when the session db is not in-memory.

nsklikas commented:
> I actually think that mapping scopes to claims and specifying claims using the claims parameter are alternative ways of requesting claims. They should NOT be used together. If they are, whatever is defined using the claims parameter overrides the scope->claims transformation.

As I posted above:
`If present, the listed Claims are being requested to be added to any Claims that are being requested using scope values.`

This means that claims and scopes can be used to together. The current behavior seems counter-intuitive, but if you think that this should be the way forward we can simply override it in our implementation.

rohe commented:
OK, I can live with the claim set returned being the union of 
1. scope parameter values->claims mapping, 
2. additional claims that are defined in configuration 
3. and/or in the client DB and 
4. what is defined in the claims parameter.

But I will insist on what's in the claims parameter overrides what's defined elsewhere for the same claim.

For example, if we have:
scope="openid email" and
claims={"userinfo": {"email": {"value": ["foo@example.com"]}}}

and for the user in question we only have the email address bar@example.com. 
Then bar@example.com will  NOT be returned.
I think this is the proper way to adhere to GDPR data minimisation.

nsklikas commented:
I agree with this approach.

rohe commented:
Excellent! Will implement it tomorrow.

nsklikas commented on src/oidcendpoint/endpoint_context.py:
```diff
@@ -144,6 +136,7 @@ def __init__(
         self.login_hint2acrs = None
         self.userinfo = None
         self.scope2claims = SCOPE2CLAIMS
+        self.token_usage_rules = {}
```
Is there any way to initialize this through configuration, on initialization? 

I can't find out how to define this other that setting after the endpoint_context has been initialized:
```
ctx = EndpointContext()
ctx.token_usage_rules = some_dict
```

rohe replied:
Ah, there is a disconnect here. I'll fix that and will add a couple of tests.
You can find token_usage_rules in the configuration of test_35_... and test_50_... but not used :-(

rohe replied:
Hmm, have been a bit unclear in my thinking here. Token usage rules are part of the authorization handling so it should not be on the top-level of endpoint_context. Look at test_35_.. authz/kwargs/grant_config/ that is where it should be.

rohe committed:
Cleaned up handling of token usage rules.

angelakis commented on src/oidcendpoint/session/grant.py:
```diff
@@ -0,0 +1,337 @@
+            "token_map": {k: ".".join([v.__module__, v.__name__]) for k, v in
+                          self.token_map.items()}
+        }
+        if isinstance(self.authorization_request, Message):
```
Could we add the `authorization_request` and `authentication_event` to the dict above? So that it works when these objects are dicts too.

rohe replied:
I've taken care of the 2 cases (Message type or dictionary).

nsklikas commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -83,26 +115,89 @@ def _access_token(self, req, **kwargs):
+
+        _response = {
+            "token_type": "Bearer",
+            "expires_in": 900,
```
This is still hardcoded, this should be the access token's lifetime

rohe replied:
I've changed this.

nsklikas commented on src/oidcendpoint/oauth2/authorization.py:
```diff
@@ -100,27 +287,51 @@ class Authorization(Endpoint):
+            usage_rules=usage_rules
+        )
+
+        _exp_in = usage_rules.get("expires_in")
+        if isinstance(_exp_in, str):
+            _exp_in = int(_exp_in)
+        if _exp_in:
+            token.expires_at = utc_time_sans_frac() + _exp_in
```
Shouldn't this be done in `grant.mint_token`?

rohe replied:
mint_token() only sets expires_at and that is because expires_at is relative to the minter's time and is absolute.
When a token is presented to the introspection endpoint, expires_at is used to see if the token has timed-out.
One could also use expires_in + issued_at but I choose to go this way.

expires_in on the other hand is relative which means you have to know when it was minted or when you received it to be able to figure out if the token is too old.

rohe committed:
Accept dicts instead of Messages for authorization_request and authentication_event.

rohe committed:
Changed the use union instead of intersection between gathered claims (claims->scope, config, client config) and those specified using the request parameter.

nsklikas commented on src/oidcendpoint/authz/__init__.py:
```diff
@@ -1,63 +1,101 @@
+            grant.resources = resources
+
+        # This is where user consent should be handled
+        for interface in ["userinfo", "introspection", "id_token", "token"]:
```
I think we should rename `token` to `access_token`, for clarity and backwards compatibility. If we leave it like this then any user, that uses claims_per_client for the access token, will have to change all his client metadata from `access_token_claims` to `token_claims`.

I don't see any reason to change the name

rohe replied:
OK

rohe replied:
There are other places where there might be reason to change 'token' to 'access_token'.

rohe committed:
No static data.

nsklikas commented on src/oidcendpoint/authz/__init__.py:
```diff
@@ -1,63 +1,101 @@
-    def __init__(self, endpoint_context, permission="implicit"):
-        AuthzHandling.__init__(self, endpoint_context)
-        self.permission = permission
+        scope = request.get("scope")
```
`request.get("scope") ` may return a space separated list, we should split the scopes before passing them to args

rohe replied:
This is why I've always questioned why scope was this stupid space separated 'list'.
But the powers that be wouldn't listen to me.
Anyway if request is a Message instance you will **always** get a list. If request is a dict then all bets are off.

nsklikas commented:
An issue that I have is that `token_coop.py` was deleted, without being "merged" with `token.py`. This means that some of the changes we had done in token_coop were lost. E.g. It seems like how/when a refresh token is issued has changed

nsklikas commented on src/oidcendpoint/session/claims.py:
```diff
@@ -0,0 +1,181 @@
+
+        # This will add claims that has not be added before and
+        # set filters on those claims that also appears in one of the sources above
+        for k,v in request_claims.items():
+            claims[k] = v
```
```python
claims.update(request_claims)
```

rohe replied:
Of course

rohe commented:
Regarding token_coop.py I could go back and try to find out what was left behind and/or you could tell me what's missing in the new token.py .

rohe commented:
I'm rewriting token.py basing it on token_coop.py instead of the old token.py

rohe committed:
Changed to claims.update.

rohe committed:
Added more info to authorization response that contains a token.

rohe committed:
Not hard coded.

rohe committed:
token -> access_token

rohe committed:
Changed token.py to be based on the old token_coop.py instead of the old token.py .

rohe committed:
Fixed tests.

rohe committed:
A revoked token is caught when parsing a request not when processing it. Saves some cycles.

rohe committed:
Cleaning up.

nsklikas commented on src/oidcendpoint/oidc/userinfo.py:
```diff
@@ -34,15 +36,16 @@ class UserInfo(Endpoint):
 
     def __init__(self, endpoint_context, **kwargs):
         Endpoint.__init__(self, endpoint_context, **kwargs)
```
Shouldn't `add_claims_by_scope` default to True for userinfo?
Something like this?
```python
    def __init__(self, endpoint_context, add_claims_by_scope=True, **kwargs):
        Endpoint.__init__(
            self,
            endpoint_context,
            add_claims_by_scope=add_claims_by_scope,
            **kwargs,
        )
```

rohe replied:
Probably

nsklikas commented on src/oidcendpoint/session/claims.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Scopes can in some cases equate to set of claims, is that used here ?
+        if module and module.kwargs.get("add_claims_by_scope"):
+            _supported = self.endpoint_context.provider_info.get("scopes_supported", [])
+            if _supported:
+                _scopes = set(_supported).intersection(set(scopes))
+            else:
+                _scopes = scopes
```
Shouldn't this use the client's allowed scopes?
```python
            _scopes = self.endpoint_context.scopes_handler.filter_scopes(
                client_id, self.endpoint_context, scopes
            )
```

Where `filter_scopes` is:
```python
class Scopes:
    ...
    def filter_scopes(self, client_id, endpoint_context, scopes):
        allowed_scopes = self.allowed_scopes(client_id, endpoint_context)
        return [s for s in scopes if s in allowed_scopes]
```

rohe replied:
get_claims() has scopes as one parameter. Are you proposing that we run that through filter_scopes before intersecting it with what the provider allows ?

nsklikas commented on src/oidcendpoint/session/grant.py:
```diff
@@ -0,0 +1,355 @@
+            "issued_token": [t.to_json() for t in self.issued_token],
+            "id": self.id,
+            "token_map": {k: ".".join([v.__module__, v.__name__]) for k, v in
+                          self.token_map.items()}
```
We need to add the `usage_rules` to the json

rohe replied:
Yepp.

nsklikas commented on src/oidcendpoint/session/grant.py:
```diff
@@ -0,0 +1,355 @@
+    if not _usage:
+        _usage = DEFAULT_USAGE[token_type]
+
+    _cinfo = endpoint_context.cdb.get(client_id, {})
+    if "token_usage_rules" in _cinfo:
+        try:
+            _usage.update(_cinfo["token_usage_rules"][token_type])
+        except KeyError:
+            pass
```
Isn't this done from `usage_rules_for`?

rohe replied:
Looks like it.

nsklikas commented on src/oidcendpoint/session/grant.py:
```diff
@@ -0,0 +1,355 @@
+    :return: Usage specification
+    """
+
+    _usage = endpoint_context.authz.usage_rules_for(token_type)
```
```python
     _usage = endpoint_context.authz.usage_rules_for(client_id, token_type)
```

rohe replied:
Hmm, tests should have caught this. Have to write more.

nsklikas commented on src/oidcendpoint/session/manager.py:
```diff
@@ -0,0 +1,429 @@
+        except AttributeError:
+            sector_identifier = ""
+
+        grant = Grant(authorization_request=auth_req,
+                      authentication_event=authn_event,
+                      sub=self.sub_func[sub_type](
+                          user_id, salt=self.salt,
+                          sector_identifier=sector_identifier),
+                      usage_rules=token_usage_rules
+                      )
```
Add scopes to the grant.

rohe replied:
So, my thinking was that you first get a grant instance with a basic set of attributes (scopes and claims are not among those) and then you add to that after you've dealt with user consent and what ever other policy processes you have to go through.

nsklikas replied:
I see, it makes sense. I guess you could add the option to add scopes to grant in creation (through a scopes param in `create_session` and `create_grant` which would default to False). My thinking is that when using a database each call is a query so I would want to make as few as possible when creating a grant.

Since in some use cases you already know the scopes/claims you want to release when the user logs in, it makes sense to provide this option

rohe replied:
I can definitely buy the optimisation argument and having a scopes parameter in create_session/create_grant would not be the end of the world.
I agree that in some use cases you know what you want to release but then there is user consent and you may have to change the grant anyway. Granted it may not happen every time so we'll save occasional calls to the database.
OK, I'm sold I'll add the parameters.

nsklikas commented:
It seems to me that it could be problem that we only store scopes in the Grant.

The scopes of a token might be different than those of the grant. Both `token_exchange` and `refresh_token` grants can alter the scopes. `Grant` seems like a proper place to store the original scopes, but we need to know the scopes of each token as well. If the token is a a JWT then that is simple, but in the case where the token is a random string we need to store somewhere the scopes associated with the token.

Perhaps we should store scopes in the tokens as well as the grant?

rohe commented:
If you look at oidcendpoint.session.token.Token you can see that scope is in there so we don't just have scopes in Grants.
There might not be a test for setting specific scopes in a token but that was definitely in the plans.
What's there is code that finds out what scopes are valid for a token by first looking in the token then at the grant that lies behind the token. If there are scopes specified in the token the processing stops there. If there is no scopes defined for the token the once added to the grant is used.

nsklikas commented:
Very well, I'm sorry then. There are a lot of changes so it's easy to get lost.

nsklikas commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -54,21 +102,18 @@ def _access_token(self, req, **kwargs):
-            return self.error_cls(
-                error="invalid_grant", error_description="Session is revoked"
-            )
+        if "state" in req:
+            # verify that state in this request is the same as the one in the
+            # authorization request
+            if _authn_req["state"] != req["state"]:
+                logger.error("State value mismatch")
+                return self.error_cls(error="invalid_request")
```
This must be removed, there is no `state` in the token request

rohe replied:
OK

rohe commented:
I'm really appreciating your review. It's excellent !!

rohe committed:
If scopes aren't set then matching them against what's allowed doesn't make sense.

rohe committed:
Better name on reference.
Added usage_rules to json payload.
Missing parameter client_id added.
Removed code that was already done elsewhere.

rohe committed:
Added scopes parameter to create_grant and create_session.

rohe committed:
Added method filer_scopes().

rohe committed:
Test for get_usage_rules

rohe committed:
Don't depend on order.

rohe committed:
Don't depend on order.

rohe committed:
Removed code that didn't make sense since 'state' is not required in token request.

rohe committed:
Made add_claims_by_scope being the default.

peppelinux commented on src/oidcendpoint/endpoint_context.py:
```diff
@@ -311,9 +298,9 @@ def do_login_hint_lookup(self):
         _conf = self.conf.get("userinfo")
         if _conf:
-            if self.sdb:
+            if self.session_manager:
```
yes "session_manager" is more readable than sdb! :ok_hand:

nsklikas commented on src/oidcendpoint/oauth2/authorization.py:
```diff
@@ -507,10 +794,8 @@ def post_authentication(self, user, request, sid, **kwargs):
-            sub=user,
-            sid=sid,
+            sid=session_id,
             state=request["state"],
```
```python
request.get("state")
```
state is optional

peppelinux replied:
state is recommendend in oidc ...

rohe replied:
But not required. Hence if it's there in the request we add it to the cookie info otherwise it will just be ignored.

nsklikas commented on src/oidcendpoint/oidc/add_on/pkce.py:
```diff
@@ -124,21 +121,7 @@ def post_token_parse(request, client_id, endpoint_context, **kwargs):
 
 
 def add_pkce_support(endpoint, **kwargs):
-    authn_endpoint = endpoint.get("authorization")
-    if authn_endpoint is None:
-        LOGGER.warning(
-            "No authorization endpoint found, skipping PKCE configuration"
-        )
-        return
-
-    token_endpoint = endpoint.get("token")
-    if token_endpoint is None:
-        LOGGER.warning(
-            "No token endpoint found, skipping PKCE configuration"
-        )
-        return
-
-    authn_endpoint.post_parse_request.append(post_authn_parse)
```
Why was this removed?

peppelinux replied:
probably a better path for this module would be `src/oidcendpoint/oauth2/add_on/pkce.py` as it belongs to OAuth2 and can be used even in oidc

rohe replied:
Right!

angelakis commented on src/oidcendpoint/oidc/token.py:
```diff
@@ -111,50 +254,132 @@ def _access_token(self, req, **kwargs):
+                    grant_type_options in (None, True)
+                    and grant_type in HELPER_BY_GRANT_TYPE
+            ):
+                self.helper[grant_type] = HELPER_BY_GRANT_TYPE[grant_type]
```
The class will need to be initialized ->
`self.helper[grant_type] = HELPER_BY_GRANT_TYPE[grant_type](self)`

angelakis commented on tests/test_35_oidc_token_endpoint.py:
```diff
@@ -149,169 +173,125 @@ def create_endpoint(self, conf):
-            },
-        },
-        {
-            "authorization_code": True,  # Both True and None end up using the defaults
```
These tests were removed and would catch the bug I reviewd above relating to the initialization of the helper endpoint classes.

angelakis commented on src/oidcendpoint/session/database.py:
```diff
@@ -0,0 +1,147 @@
+            return user_info
+        else:
+            if client_id not in user_info['subordinate']:
+                raise ValueError('No session from that client for that user')
```
`raise NoSuchGrant` (or even `NoSuchClientSession`)

angelakis commented on src/oidcendpoint/oidc/add_on/pkce.py:
```diff
@@ -96,11 +90,14 @@ def post_token_parse(request, client_id, endpoint_context, **kwargs):
         return request
```
This is probably not strictly related to this PR but could be fixed here anyway. A check should be added here for the grant type because we don't want to check for `code` in a refresh token grant or a token exchange one.

Or we could check for the request instance like this (maybe strictly checking for the grant type to be authorization_code is cleaner tho):
```
if isinstance(
    request,
    (AuthorizationErrorResponse, RefreshAccessTokenRequest, TokenExchangeRequest),
):

```

nsklikas commented:
This PR has been open for long enough. We should close it but there are still some open issues

If @rohe is ok with it I can push the fixes required and you can review them so that we can finally close this PR.

rohe commented:
Let's !!

I've been side tracked by the persistent storage handling.
Have done the RP side. almost done with the OP side when I hit a snag earlier today :-(

nsklikas committed:
Properly handle dict

nsklikas committed:
Allow grant to infer token_handler

The access_token handler was used to mint auth codes

nsklikas committed:
State is optional

nsklikas committed:
Check if token and authorization exist on pkce

nsklikas committed:
Handle unknown token

nsklikas committed:
Don't return state in token response

nsklikas committed:
Get token from grant instead of manager

nsklikas committed:
Keep backwards compatibility with client claims

nsklikas committed:
Filter scopes per client

angelakis committed:
Merge pull request #106 from nsklikas/new_session_handling

New session handling

nsklikas committed:
Fix bug, default token_handler to None

nsklikas committed:
Merge pull request #107 from nsklikas/new_session_handling

Fix bug, default token_handler to None

angelakis committed:
Replace find_token with get_token to avoid DB call

We already have the grant in most cases so we can fetch the token from
inside it instead of querying the DB again.

rohe committed:
Merge pull request #108 from angelakis/new-session-handling-fixes

WIP: Make more fixes to the new_session_handling before it gets merged
